### PR TITLE
Inject random errors in S3BlobStoreRepositoryIT

### PR DESCRIPTION
**Pull request by tlrx**
This pull request modifies the HTTP server used in `S3BlobStoreRepositoryTests` so that it randomly returns server errors for any type of request executed by the SDK client. It is now possible to verify that the repository tests are successfully completed even if one or more errors were returned by the S3 service in response of a blob upload, a blob deletion or a object listing request etc.

Because injecting errors forces the SDK client to retry requests, the test limits the maximum errors to send in response for each request at 3 retries. ~~The PR also renames the test class so that it now have the IT suffix.~~

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1 (unrelated REST failure)

File: plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobStoreRepositoryTests.java [hunk 1]
```diff
@@ -229,4 +238,50 @@ public void handle(final HttpExchange exchange) throws IOException {
...
+    /**
+     * HTTP handler that injects random S3 service errors
+     *
+     * Note: it is not a good idea to allow this handler to simulates too many errors as it would
+     * slow down the test suite and/or could trigger SDK client request throttling (and request
+     * would fail before reaching the max retry attempts - this can be mitigated by disabling
+     * {@link S3ClientSettings#USE_THROTTLE_RETRIES_SETTING})
+     */
+    @SuppressForbidden(reason = "this test uses a HttpServer to emulate an S3 endpoint")
+    private class ErroneousHttpHandler implements HttpHandler {
```
Comment (original-brownbear) [inline plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobStoreRepositoryTests.java]: NIT: can be `static`

File: plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobStoreRepositoryTests.java [hunk 2]
```diff
@@ -229,4 +238,50 @@ public void handle(final HttpExchange exchange) throws IOException {
             }
         }
     }
+
+    /**
+     * HTTP handler that injects random S3 service errors
+     *
+     * Note: it is not a good idea to allow this handler to simulates too many errors as it would
```
Comment (original-brownbear) [inline plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobStoreRepositoryTests.java]: NIT: `-s` => `simulates` -> `simulate`

File: plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobStoreRepositoryTests.java [hunk 3]
```diff
@@ -229,4 +238,50 @@ public void handle(final HttpExchange exchange) throws IOException {
...
+            this.maxErrorsPerRequest = maxErrorsPerRequest;
+            assert maxErrorsPerRequest > 1;
+        }
+
+        @Override
+        public void handle(final HttpExchange exchange) throws IOException {
+            final String requestId = exchange.getRequestHeaders().getFirst(AmazonHttpClient.HEADER_SDK_TRANSACTION_ID);
+            assert Strings.hasText(requestId);
+
+            final int count = requests.computeIfAbsent(requestId, (req) -> new AtomicInteger(0)).incrementAndGet();
```
Comment (original-brownbear) [inline plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobStoreRepositoryTests.java]: NIT: brackets around `req` are redundant

**Review by original-brownbear — APPROVED** [review]
LGTM :)

**Comment by tlrx** [comment]
Thanks @original-brownbear !

**Pull request merged**